### PR TITLE
Adds test cases for edge cases found to be lacking test coverage

### PR DESCRIPTION
**Pull request by thorbjoernl**
## Change Summary

Adds additional test cases for parts of the code base that previously lacked test coverage. This includes testing reading and writing with different AccessTypes. Includes changes from #60 so should be merged after that one.

Additional tests introduced:
- Reading and writing json directly (access_type=JSON_STR).
- Add tests for edge cases found using mutation testing.

## Related issue number
NA

## Checklist

* [x] Start with a draft-PR
* [x] The PR title is a good summary of the changes
* [x] PR is set to AeroTools and a tentative milestone
* [x] Documentation reflects the changes where applicable
* [x] Tests for the changes exist where applicable
* [x] Tests pass locally
* [x] Tests pass on CI
* [x] At least 1 reviewer is selected
* [x] Make PR ready to review

### CODE REVIEW

File: src/aerovaldb/jsondb/jsonfiledb.py [hunk 1]
```diff
@@ -241,9 +148,9 @@ async def _get_version(self, project: str, experiment: str) -> Version:
                 version = Version("0.0.1")
             finally:
                 return version
-        except simplejson.errors.JSONDecodeError:
-            # Work around for https://github.com/metno/aerovaldb/issues/28
-            return Version("0.14.0")
+        # except simplejson.JSONDecodeError:
+        #    # Work around for https://github.com/metno/aerovaldb/issues/28
+        #    return Version("0.14.0")
```
Comment (heikoklein) [inline src/aerovaldb/jsondb/jsonfiledb.py:153]: dead code?

File: src/aerovaldb/jsondb/jsonfiledb.py [hunk 2]
```diff
@@ -317,52 +184,64 @@ async def _get(
...
-                if isinstance(data, str):
-                    data = simplejson.loads(data, allow_nan=True)
+        if not os.path.exists(file_path):
+            if default is None or access_type == AccessType.FILE_PATH:
+                raise FileNotFoundError(f"File {file_path} does not exist.")
+            return default
+
+        # No filtered.
+        if filter_func is None:
+            if access_type == AccessType.FILE_PATH:
```
Comment (heikoklein) [inline src/aerovaldb/jsondb/jsonfiledb.py:219]: comment doesn't add anything useful

**Review by heikoklein — CHANGES_REQUESTED** [review]
With the added code from #60 this is currently not possible to review. Should be a draft.

File: src/aerovaldb/sqlitedb/sqlitedb.py [hunk 3]
```diff
@@ -436,7 +436,10 @@ async def _get(self, route, route_args, *args, **kwargs):
         raise UnsupportedOperation
 
     async def _put(self, obj, route, route_args, *args, **kwargs):
-        assert len(args) == 0
+        if len(args) > 0:
+            raise UnusedArguments(
+                f"Unexpected positional arguments {args}. Jsondb does not use additional positional arguments currently."
+            )
```
Comment (heikoklein) [inline src/aerovaldb/sqlitedb/sqlitedb.py:442]: Not sure why this is here? A user should transparently switch between one implementation and another, so *args and *kwargs can only be optional arguments. They should be silently ignored by other implementations.

Or, if we don't have any use-case for *args, maybe remove it completely?

Comment (heikoklein) [reply src/aerovaldb/sqlitedb/sqlitedb.py:442]: And the code seems to be a bad duplicate, since it mentions Jsondb in the sqlitedb implementation.

File: src/aerovaldb/sqlitedb/sqlitedb.py [hunk 4]
```diff
@@ -616,3 +619,34 @@ def list_timeseries(
...
+            "heatmap_timeseries2",
+            "forecast",
+            "gridded_map",
+        ]:
+            cur.execute(
+                f"""
+                DELETE FROM {table} WHERE project=? AND experiment=?
+                """,
+                (project, experiment),
+            )
```
Comment (heikoklein) [inline src/aerovaldb/sqlitedb/sqlitedb.py:652]: I would expect `rm_experiment_data` to remove all experiment data, including the entry in `experiments.json`.

File: tests/test_aerovaldb.py [hunk 5]
```diff
@@ -462,3 +508,19 @@ def test_list_timeseries(testdb):
...
+def test_rm_experiment_data(tmpdb):
+    with aerovaldb.open("json_files:./tests/test-db/json") as db:
+        for i, uri in enumerate(db.list_all()):
+            data = db.get_by_uri(uri, access_type=aerovaldb.AccessType.JSON_STR)
+            tmpdb.put_by_uri(data, uri)
+
+        assert len(list(db.list_all())) == len(list(tmpdb.list_all()))
+
+        tmpdb.rm_experiment_data("project", "experiment")
+
```
Comment (heikoklein) [inline tests/test_aerovaldb.py:525]: test for experiments.json missing

File: tests/utils/test_string_utils.py [hunk 6]
```diff
@@ -32,6 +32,7 @@ def test_str_to_bool_exception_2():
 
 
 def test_str_to_bool_default():
+    assert not str_to_bool("blah")
```
Comment (heikoklein) [inline tests/utils/test_string_utils.py:35]: I expect a ValueError here, not a None returned, similar to float:
```
>>> float("abc")
ValueError: could not convert string to float: 'abc'
```

Example: What happens if x = "blah" in the following case:
```python
if str_to_bool(x):
   print("True")
else:
   print("False")
```

Comment (heikoklein) [reply tests/utils/test_string_utils.py:35]: And looking even closer: a default of True does not make sense. The only case where the exception in line 25 is raised is for None.

**Review by heikoklein — CHANGES_REQUESTED** [review]
I see some inconsistencies in the implementation. Please check or comment.

File: src/aerovaldb/sqlitedb/sqlitedb.py (hunk 4 above)
Comment (thorbjoernl) [reply src/aerovaldb/sqlitedb/sqlitedb.py:652]: I think that as long as aerovaldb isn't responsible for writing the experiments file (which it doesn't currently do), it shouldn't remove anything from this file either. There is a whole bunch of file-synchronization code in pyaerocom ExperimentOutput that keeps files in sync, and I don't think splitting that logic into two repos is an improvement over the current state.

That being said. I like the idea of having experiments be a 'virtual' query which is generated based on the other data files present in the db.

File: src/aerovaldb/sqlitedb/sqlitedb.py (hunk 3 above)
Comment (thorbjoernl) [reply src/aerovaldb/sqlitedb/sqlitedb.py:442]: Agreed. I did this to mask confusing error messages if too many arguments are provided.

I think it will be necessary to still raise an exception, to provide a useful error message, but it should probably be in the get/put decorators.

File: tests/utils/test_string_utils.py (hunk 6 above)
Comment (thorbjoernl) [reply tests/utils/test_string_utils.py:35]: I was going for matching the pattern seen in dict.get() where you can provide a default to be returned instead of raising KeyError. The strict option is unnecessary so I will change it to default=None instead to match the dict implementation. So:

```
str_to_bool("blah") # ValueError
str_to_bool("blah", default=False) # False
```

Does this sound ok?

Comment (heikoklein) [reply tests/utils/test_string_utils.py:35]: `dict.get()` is already the specialized code, whereas `dict[]` is the normal case. I like your example: a value-error without explicit `default` and the explicit default otherwise.

File: src/aerovaldb/utils/string_utils.py [hunk 7]
```diff
@@ -4,21 +4,22 @@
...
     Supports 1/0, true/false, t/f, yes/no, y/n (Case insensitive).
 
     :param value : The string value to be converted.
-    :param strict : If true, ValueError is raised if an unrecognized value is encountered,
-    otherwise default is returned.
-    :param default : Returned if strict is disabled, for unrecognized values.
+    :param default : If None, a string value not explicitly matching one of the
+        above values for true/false will raise a ValueError. Otherwise this default
+        will be returned instead.
+
```
Comment (heikoklein) [inline src/aerovaldb/utils/string_utils.py:18]: You document here foremost an internal value (None) which the user is not supposed to set. I suggest: 
```
Optional default return value (True/False) if the string value doesn't match any supported value. If not set, or set to None, a ValueError is raised in these cases.
```

File: src/aerovaldb/utils/string_utils.py [hunk 8]
```diff
@@ -4,21 +4,22 @@
...
     :raises ValueError
         If value is not a string
     :raises ValueError
-        If strict is true, and value does not match expected values.
+        If default is None, and value does not match one of the above templates.
```
Comment (heikoklein) [inline src/aerovaldb/utils/string_utils.py:22]:
```
Raised on unsupported input value. Suppressed if default is set to True/False.
```

File: src/aerovaldb/utils/string_utils.py [hunk 9]
```diff
@@ -29,10 +30,10 @@ def str_to_bool(value: str, /, strict: bool = False, default: bool = False) -> b
     if value.lower() in ["0", "false", "f", "no", "n"]:
         return False
 
-    if not strict:
+    if default is not None:
         return default
 
-    raise ValueError("Unexpected string '{value}'")
+    raise ValueError("Could not convert string to float: '{value}'")
```
Comment (heikoklein) [inline src/aerovaldb/utils/string_utils.py:36]: to float -> to bool

**Review by heikoklein — APPROVED** [review]
Looks good, just some suggestions on improving documentation/user-feedback.

**Review by heikoklein — APPROVED** [review]
